embulk-input-jdbc/src/main/java/org/embulk/input/jdbc/getter/ArrayColumnGetter.java

```diff
@@ -0,0 +1,103 @@
+import org.embulk.spi.json.JsonParser;
+import org.embulk.spi.type.Type;
+import org.msgpack.value.Value;
+
+import java.sql.Array;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+
+import java.sql.Types;
+
+public class ArrayColumnGetter
+        extends AbstractColumnGetter
+{
+    protected Array value;
+
+    protected final JsonParser jsonParser = new JsonParser();
+
+    public ArrayColumnGetter(PageBuilder to, Type toType)
+    {
+        super(to, toType);
+    }
+
+    @Override
+    protected void fetch(ResultSet from, int fromIndex) throws SQLException
+    {
+        value = from.getArray(fromIndex);
+    }
+
+    private ArrayNode buildJsonArray(Object[] elements)
+            throws SQLException
+    {
+        ArrayNode arrayNode = jsonNodeFactory.arrayNode();
+        for (Object v : elements) {
+            if (v == null) {
+                arrayNode.add(NullNode.getInstance());
+                continue;
+            }
+            if (v.getClass().isArray()) {
+                arrayNode.add(buildJsonArray((Object[]) v));
+            }
+            else {
+                switch (value.getBaseType()) {
+                    case Types.TINYINT:
+                    case Types.SMALLINT:
+                    case Types.INTEGER:
+                    case Types.BIGINT:
+                        arrayNode.add(jsonNodeFactory.numberNode(((Number) v).longValue()));
+                        break;
+                    case Types.FLOAT:
+                    case Types.REAL:
+                    case Types.DOUBLE:
+                        arrayNode.add(jsonNodeFactory.numberNode(((Number) v).doubleValue()));
+                        break;
+                    case Types.BOOLEAN:
+                    case Types.BIT:  // JDBC BIT is boolean, unlike SQL-92
+                        arrayNode.add(jsonNodeFactory.booleanNode((Boolean) v));
+                        break;
+                    case Types.CHAR:
+                    case Types.VARCHAR:
+                    case Types.LONGVARCHAR:
+                    case Types.CLOB:
+                    case Types.NCHAR:
+                    case Types.NVARCHAR:
+                    case Types.LONGNVARCHAR:
+                        arrayNode.add(jsonNodeFactory.textNode((String) v));
+                        break;
+                }
+            }
+        }
+        return arrayNode;
+    }
+
+    @Override
+    protected Type getDefaultToType()
+    {
+        return org.embulk.spi.type.Types.STRING;
+    }
+
+    @Override
+    public void jsonColumn(Column column)
+    {
+        try {
+            Value v = jsonParser.parse(buildJsonArray((Object[]) value.getArray()).toString());
+            to.setJson(column, v);
+        }
+        catch (JsonParseException | SQLException | ClassCastException e) {
+            super.jsonColumn(column);
+        }
+    }
+
+    @Override
+    public void stringColumn(Column column)
+    {
+        to.setString(column, value.toString());
```

Comment:
java.sql.Array is interface and doesn't implement toString method.
We should implement the code converting Array to String by hand, otherwise the result will vary depends on DBMS.

Reply:
Yes, because the implementation is kinda specific for Postgres, I refactored the code from `embulk-input-jdbc` to `embulk-input-postgresql`.